feat: update the Kotlin evolution principles, Kotlin roadmap, add language design proposals (#4385)
JetBrains/kotlin-web-site: docs/topics/kotlin-language-features-and-proposals.md

[//]: # (title: Kotlin language features and proposals)
[//]: # (description: Learn about the lifecycle of Kotlin features. 
The page contains the full list of Kotlin language features and design proposals.)

JetBrains evolves the Kotlin language according to the [Kotlin language evolution principles](kotlin-evolution-principles.md),
guided by pragmatic design. 
You can find the list of language feature proposals and their statuses.

> ## Kotlin language features lifecycle

Reviewer: My proposal is to put this section somewhere around `docs/kotlin-evolution-principles.html#language-and-tooling-releases`, so that a user won't have to scroll all the way down every time

The Kotlin language feature can have one of the following statuses:

* **Exploration and design**. We are considering the introduction of a new feature to the language.
  This involves discussing how it would integrate with existing features, gathering use cases, 
  and assessing its potential impact.
  We need feedback from users on the problems this feature would solve and the use cases it addresses.

>   Estimating how often these use cases and problems occur would also be beneficial.

Reviewer 2: Do you mean that it would be beneficial to receive estimates from users how often such use cases and problems occur? It's not clear to me 🤔

Author: Hmm, let me figure out how to rephrase that.

  Typically, ideas are documented as YouTrack issues, where the discussion continues.

* **KEEP discussion**. We are fairly certain that the feature should be added to the language.
  We aim to specify motivation, use-cases, design, and other important details in a document called _KEEP_. 
  We expect feedback from users to focus on discussing all the information provided in the KEEP.

* **In preview**. A feature prototype is ready, and you can enable it using a feature-specific compiler option.
  We seek feedback on your experience with the feature, including how easily it integrates into your codebase,
  how it interacts with existing code, and any IDE support issues or suggestions.
  The feature's design may change significantly, or it could be completely revoked based on feedback. When feature is 
  _in preview_, it has one of the [stability levels](components-stability.md#stability-levels-explained).

* **Stable**. The language feature is now a first-class citizen in the Kotlin language.
  We guarantee its backward compatibility and tooling support.

* **Revoked**. We have revoked the proposal and will not implement the feature in the Kotlin language.

>   We may revoke a feature that is _in preview_ if it does not fit the language.

Reviewer 2: ```suggestion
  We may revoke a feature that is _in preview_ if it is not a good fit for Kotlin.
```

## List of Kotlin language features and design proposals

> > This table contains Kotlin language features and proposals starting from 1.7.0.

Reviewer 2: ```suggestion
> This table contains Kotlin language features and proposals starting from Kotlin 1.7.0.
```

>
{type="note"}

<tabs>
<tab id="all-proposals" title="All">

<!-- <include include-id="all-proposals" src="all-proposals.xml"/> -->

<chunk id="source">
<table header-style="left">

> <!-- EXPLORATION AND DESIGN BLOCK -->

Reviewer: Plus six more KEEPs:

```
<tr filter="keep">
<td>

**KEEP discussion**

</td>
<td>

**KMP Kotlin-to-Java direct actualization**

* KEEP proposal: [kmp-kotlin-to-java-direct-actualization.md](https://github.com/Kotlin/KEEP/blob/kotlin-to-java-direct-actualization/proposals/kmp-kotlin-to-java-direct-actualization.md)
* Target issue: [KT-67202](https://youtrack.jetbrains.com/issue/KT-67202)

</td>
</tr>
```

```
<tr filter="keep">
<td>

**KEEP discussion**

</td>
<td>

**Streamline KDoc ambiguity links**

* KEEP proposal: [streamline-KDoc-ambiguity-references.md](https://github.com/Kotlin/KEEP/blob/kdoc/Streamline-KDoc-ambiguity-references/proposals/kdoc/streamline-KDoc-ambiguity-references.md)
* Target issues: [dokka/#3451](https://github.com/Kotlin/dokka/issues/3451), [dokka/#3179](https://github.com/Kotlin/dokka/issues/3179), [dokka/#3334](https://github.com/Kotlin/dokka/issues/3334)

</td>
</tr>
```

```
<tr filter="keep">
<td>

**KEEP discussion**

</td>
<td>

**Resolution of links to extensions in KDoc**

* KEEP proposal: [links-to-extensions.md](https://github.com/Kotlin/KEEP/blob/kdoc/extension-links/proposals/kdoc/links-to-extensions.md)
* Target issue: [dokka/#3555](https://github.com/Kotlin/dokka/issues/3555)

</td>
</tr>
```

```
<tr filter="keep">
<td>

**KEEP discussion**

</td>
<td>

**Uuid**

* KEEP proposal: [uuid.md](https://github.com/Kotlin/KEEP/blob/uuid/proposals/stdlib/uuid.md)
* Target issue: [KT-31880](https://youtrack.jetbrains.com/issue/KT-31880)

</td>
</tr>
```

```
<tr filter="keep">
<td>

**KEEP discussion**

</td>
<td>

** Improve resolution using expected type**

* KEEP proposal: [improved-resolution-expected-type.md](https://github.com/Kotlin/KEEP/blob/improved-resolution-expected-type/proposals/improved-resolution-expected-type.md)
* Target issue: [KT-16768](https://youtrack.jetbrains.com/issue/KT-16768)

</td>
</tr>
```

<tr filter="exploration-and-design">
<td>

**Exploration and design**

</td>
<td>

**Kotlin statics and static extensions**

* KEEP proposal: [statics.md](https://github.com/Kotlin/KEEP/blob/statics/proposals/statics.md)
* YouTrack issue: [KT-11968](https://youtrack.jetbrains.com/issue/KT-11968)

</td>
</tr>

<tr filter="exploration-and-design">
<td>

**Exploration and design**

</td>
<td>

**Collection literals**

> * KEEP proposal: [collection-literals.md](https://github.com/KyLeggiero/KEEP/blob/collection-literals/proposals/collection-literals.md)

Reviewer: Should be "Not defined" as this proposal was closed a few years ago

* YouTrack issue: [KT-43871](https://youtrack.jetbrains.com/issue/KT-43871)

</td>
</tr>

<tr filter="exploration-and-design">
<td>

**Exploration and design**

</td>
<td>

**Union types for errors and exceptions**

* KEEP proposal: Not defined

> * YouTrack issue: [KT-7128](https://youtrack.jetbrains.com/issue/KT-7128)

Reviewer: Should be [KT-68296](https://youtrack.jetbrains.com/issue/KT-68296)

</td>
</tr>

<tr filter="exploration-and-design">
<td>

**Exploration and design**

</td>
<td>

**Name-based destructuring**

* KEEP proposal: Not defined
* YouTrack issue: [KT-19627](https://youtrack.jetbrains.com/issue/KT-19627)

</td>
</tr>

<tr filter="exploration-and-design">
<td>

**Exploration and design**

</td>
<td>

> **Support immutability (multifield value classes)**

Reviewer: Let's drop "(multifield value classes)" part, it's not only about value classes

> * KEEP proposal: Not defined

Reviewer: Let's add this: KEEP notes: [immutability](https://github.com/Kotlin/KEEP/blob/master/notes/value-classes.md#immutability-and-value-classes)

<!-- KEEP DISCUSSION BLOCK -->

<tr filter="keep">
<td>

**KEEP discussion**

</td>
<td>

**Explicit backing fields: both `public` and `private` type for the same property**

* KEEP proposal: [explicit-backing-fields.md](https://github.com/Kotlin/KEEP/blob/explicit-backing-fields-re/proposals/explicit-backing-fields.md)
* YouTrack issue: [KT-14663](https://youtrack.jetbrains.com/issue/KT-14663)
* Target version: Not defined

</td>
</tr>

<tr filter="keep">
<td>

**KEEP discussion**

</td>
<td>

**Context parameters: support for context-dependent declarations**

* KEEP proposal: [context-parameters.md](https://github.com/Kotlin/KEEP/blob/context-parameters/proposals/context-parameters.md)
* YouTrack issue: [KT-14663](https://youtrack.jetbrains.com/issue/KT-10468)
* Target version: Not defined

</td>
</tr>

<tr filter="keep">
<td>

**KEEP discussion**

</td>
<td>

**Guard conditions in when-with-subject**

* KEEP proposal: [guards.md](https://github.com/Kotlin/KEEP/blob/guards/proposals/guards.md)
* YouTrack issue: [KT-13626](https://youtrack.jetbrains.com/issue/KT-13626)
* Target version: 2.1.0

</td>
</tr>
<tr filter="keep">
<td>

**KEEP discussion**

</td>
<td>

> **Improved exhaustiveness checking in the `when` block**

Reviewer: Let's drop this part completely, we're not going to work on it right now
